File: ios/Plugin/Adapters/RawTcpAdapter.swift
import Foundation
import UIKit

/// Adapter filet de sécurité réseau iOS : raster ESC/POS sur TCP brut, sans statut.
final class RawTcpAdapter: PrinterAdapter {

    let id: AdapterId = .rawTcp
    private var connections: [String: TcpTransport] = [:]
    private let lock = NSLock()

    func isAvailable() -> Bool { true }

    func discover(timeoutMs: Int, onFound: @escaping (DiscoveredPrinter) -> Void) async {}

    func canHandle(_ profile: PrinterProfile) -> Bool {
        profile.adapter == .rawTcp && (profile.transport == .wifi || profile.transport == .ethernet)
    }

    func connect(_ profile: PrinterProfile, timeoutMs: Int) async throws {
        if isConnected(profile.id) { return }
        var host = profile.address
        var port: UInt16 = 9100
        if let idx = profile.address.lastIndex(of: ":") {
            host = String(profile.address[..<idx])
            port = UInt16(profile.address[profile.address.index(after: idx)...]) ?? 9100
        }
        let t = TcpTransport(host: host, port: port)
        try await t.open(timeoutMs: timeoutMs)
        lock.lock(); connections[profile.id] = t; lock.unlock()
    }

    func isConnected(_ printerId: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return connections[printerId]?.isOpen == true
    }

    func disconnect(_ printerId: String) async {
        lock.lock(); let t = connections.removeValue(forKey: printerId); lock.unlock()
        t?.close()
    }

    func printImage(_ profile: PrinterProfile, image: UIImage, options: RenderOptions) async throws -> Int {
        lock.lock(); let t = connections[profile.id]; lock.unlock()
        guard let transport = t else { throw PrinterError(.CONNECTION_FAILED, "rawTcp non connecté") }
        let mono = try ImageProcessor.toMono(image, options: options)
        let raster = ImageProcessor.encodeEscPosRaster(mono)
        let job = EscPosCommands.buildJob(raster: raster, align: options.align, feedLines: options.feedLines, cut: options.cut, openDrawer: options.openCashDrawer)
        try await transport.write(job)
        return job.count
    }

    func getStatus(_ profile: PrinterProfile) async throws -> PrinterStatus {
        let c = isConnected(profile.id)
        return PrinterStatus(id: profile.id, connection: c ? "connected" : "disconnected", online: c, paper: "unknown", rawStatus: "rawTcp: statut non supporté")
    }
}

File: android/src/main/java/com/resto/thermalprinter/adapters/RawTcpAdapter.kt
package com.resto.thermalprinter.adapters

import android.graphics.Bitmap
import com.resto.thermalprinter.image.ImageProcessor
import com.resto.thermalprinter.model.AdapterId
import com.resto.thermalprinter.model.DiscoveredPrinter
import com.resto.thermalprinter.model.ErrorCode
import com.resto.thermalprinter.model.PrinterException
import com.resto.thermalprinter.model.PrinterProfile
import com.resto.thermalprinter.model.PrinterStatus
import com.resto.thermalprinter.model.RenderOptions
import com.resto.thermalprinter.model.Transport
import com.resto.thermalprinter.transport.TcpTransport
import java.util.concurrent.ConcurrentHashMap

/**
 * Adapter "filet de sécurité" réseau : envoie un raster ESC/POS sur un socket TCP
 * brut sans rien présumer du dialecte de statut. Utilisé quand une imprimante
 * réseau n'est identifiée par aucun SDK et n'a pas confirmé l'ESC/POS.
 *
 * Priorité la plus basse (voir priority.ts). Ne lit jamais de statut.
 */
class RawTcpAdapter : PrinterAdapter {

    override val id = AdapterId.RAW_TCP
    private val connections = ConcurrentHashMap<String, TcpTransport>()

    override fun isAvailable(): Boolean = true

    override suspend fun discover(timeoutMs: Long, onFound: (DiscoveredPrinter) -> Unit) {
        // Délégué au TcpScanner.
    }

    override fun canHandle(profile: PrinterProfile): Boolean =
        profile.adapter == AdapterId.RAW_TCP &&
            profile.transport in setOf(Transport.WIFI, Transport.ETHERNET)

    override suspend fun connect(profile: PrinterProfile, timeoutMs: Long) {
        if (isConnected(profile.id)) return
        val idx = profile.address.lastIndexOf(':')
        val host = if (idx > 0) profile.address.substring(0, idx) else profile.address
        val port = if (idx > 0) profile.address.substring(idx + 1).toIntOrNull() ?: 9100 else 9100
        val t = TcpTransport(host, port)
        t.open(timeoutMs)
        connections[profile.id] = t
    }

    override fun isConnected(printerId: String): Boolean = connections[printerId]?.isOpen == true

    override suspend fun disconnect(printerId: String) {
        connections.remove(printerId)?.close()
    }

    override suspend fun printBitmap(profile: PrinterProfile, bitmap: Bitmap, options: RenderOptions): Int {
        val t = connections[profile.id]
            ?: throw PrinterException(ErrorCode.CONNECTION_FAILED, "rawTcp non connecté")
        val mono = ImageProcessor.toMono(bitmap, options)
        val raster = ImageProcessor.encodeEscPosRaster(mono)
        val job = EscPosCommands.buildJob(raster, options.align, options.feedLines, options.cut, options.openCashDrawer)
        t.write(job)
        return job.size
    }

    override suspend fun getStatus(profile: PrinterProfile): PrinterStatus {
        val connected = isConnected(profile.id)
        return PrinterStatus(
            id = profile.id,
            connection = if (connected) "connected" else "disconnected",
            online = connected,
            paper = "unknown",
            rawStatus = "rawTcp: statut non supporté",
        )
    }
}
